core/string/translation.cpp

```diff
@@ -56,70 +58,189 @@ void _check_for_incompatibility(const String &p_msgctxt, const String &p_msgid)
 	}
 }
 
+static Variant _write_message_key(const Translation::MessageKey &p_key) {
+	if (p_key.msgctxt.is_empty()) {
+		return p_key.msgid;
+	}
+	const Array &arr = { p_key.msgctxt, p_key.msgid };
+	return arr;
+}
+
+static bool _read_message_key(const Variant &p_raw, Translation::MessageKey &r_key) {
+	switch (p_raw.get_type()) {
+		case Variant::STRING_NAME: {
+			r_key.msgctxt = StringName();
+			r_key.msgid = p_raw;
+			return true;
+		} break;
+
+		case Variant::ARRAY: {
+			const Array &arr = p_raw;
+			r_key.msgctxt = arr[0];
+			r_key.msgid = arr[1];
+			return true;
+		} break;
+
+		default: {
+			WARN_PRINT(vformat("Unknown message key type: %s.", Variant::get_type_name(p_raw.get_type())));
+		} break;
+	}
+	return false;
+}
+
+static Vector<StringName> _read_message_value(const Variant &p_raw) {
+	switch (p_raw.get_type()) {
+		case Variant::STRING_NAME: {
+			return { p_raw };
+		} break;
+
+		case Variant::ARRAY: {
+			const Array &arr = p_raw;
+
+			Vector<StringName> msgstrs;
+			msgstrs.resize(arr.size());
+			for (int i = 0; i < arr.size(); i++) {
+				msgstrs.write[i] = arr[i];
```

Comment:
If you are doing it this way, it should be
```suggestion
			StringName *msgstrs_write = msgstrs.ptrw();
			for (int i = 0; i < arr.size(); i++) {
				msgstrs_write[i] = arr[i];
```
Alternatively just use `reserve_exact()` instead of `resize()` and append.